prototypes/ScheduledMerges.hs

```diff
@@ -163,101 +159,115 @@ mergeLastForLevel _  = MergeMidLevel
 -- | Note that the invariants rely on the fact that levelling is only used on
 -- the last level.
 --
-invariant :: forall s. Levels s -> ST s Bool
+invariant :: forall s. Levels s -> ST s ()
 invariant = go 1
   where
-    go :: Int -> [Level s] -> ST s Bool
-    go !_ []     = return True
+    go :: Int -> [Level s] -> ST s ()
+    go !_ [] = return ()
 
     go !ln (Level mr rs : ls) = do
 
       mrs <- case mr of
                SingleRun r        -> return (CompletedMerge r)
                MergingRun _ _ ref -> readSTRef ref
 
-      assert (case mr of
-                SingleRun{} -> True
-                MergingRun mp ml _ -> mergePolicyForLevel ln ls == mp
-                                   && mergeLastForLevel ls == ml)
-        assert (length rs <= 3) $
-        assert (expectedRunLengths ln rs ls) $
-        assert (expectedMergingRunLengths ln mr mrs ls) $
-        return ()
+      assertST $ case mr of
+        SingleRun{}        -> True
+        MergingRun mp ml _ -> mergePolicyForLevel ln ls == mp
+                           && mergeLastForLevel ls == ml
+      assertST $ length rs <= 3
+      expectedRunLengths ln rs ls
+      expectedMergingRunLengths ln mr mrs ls
 
       go (ln+1) ls
 
     -- All runs within a level "proper" (as opposed to the incoming runs
     -- being merged) should be of the correct size for the level.
-    expectedRunLengths :: Int -> [Run] -> [Level s] -> Bool
+    expectedRunLengths :: Int -> [Run] -> [Level s] -> ST s ()
     expectedRunLengths ln rs ls =
       case mergePolicyForLevel ln ls of
         -- Levels using levelling have only one run, and that single run is
         -- (almost) always involved in an ongoing merge. Thus there are no
         -- other "normal" runs. The exception is when a levelling run becomes
         -- too large and is promoted, in that case initially there's no merge,
         -- but it is still represented as a 'MergingRun', using 'SingleRun'.
-        MergePolicyLevelling -> null rs
-        MergePolicyTiering   -> all (\r -> tieringRunSizeToLevel r == ln) rs
+        MergePolicyLevelling -> assertST $ null rs
+        -- Runs in tiering levels usually fit that size, but they can be one
+        -- larger, if a run has been held back (creating a 5-way merge).
+        MergePolicyTiering   -> assertST $ all (\r -> tieringRunSizeToLevel r `elem` [ln, ln+1]) rs
```

Comment:
This is actually still not really true (although will hold in practice).

In the pathological case, due to holding back, all runs passed to the next level can be factor (5/4) too large, and there the same holding back can happen, leading to factor (6/4) etc., until 12 levels down a run is *two* levels too large.